lib/ktl/reassigner.rb

```diff
@@ -40,24 +40,66 @@ def load_overflow
     end
 
     def execute(reassignment)
+      reassign(reassignment)
+    end
+
+    private
+
+    JSON_MAX_SIZE = 1024**2
+
+    def reassign(reassignment)
+      if (limit)
+        @logger.info 'reassigning %d of %d partitions' % [limit, reassignment.size]
+      else
+        @logger.info 'reassigning %d partitions' % reassignment.size
+      end
+
       reassignments = split(reassignment, @limit)
-      actual_reassignment = reassignments.shift
-      if @log_assignments
-        Scala::Collection::JavaConversions.as_java_iterable(actual_reassignment).each do |pr|
-          topic_and_partition, replicas = pr.elements
+      reassignment_candidates = reassignments.shift
+      actual_reassignment = Scala::Collection::Map.empty
+      next_step_assignments = Scala::Collection::Map.empty
+      Scala::Collection::JavaConversions.as_java_iterable(reassignment_candidates).each do |pr|
+        topic_and_partition, replicas = pr.elements
+        if step1_replicas = is_two_step_operation(topic_and_partition, replicas)
+          if step1_replicas.uniq != step1_replicas
+            raise "Multiple replicas on the same broker, this should not happen... #{step1_replicas}"
+          end
+          step1_replicas = Scala::Collection::JavaConversions.as_scala_iterable(step1_replicas)
+          next_step_assignments += pr
+          actual_reassignment += Scala::Tuple.new(topic_and_partition, step1_replicas)
+          brokers = Scala::Collection::JavaConversions.as_java_iterable(step1_replicas).to_a
+          eventual_brokers = Scala::Collection::JavaConversions.as_java_iterable(replicas).to_a
+          @logger.debug "Mirroring #{topic_and_partition.topic},#{topic_and_partition.partition} to #{brokers.join(',')} for eventual transition to #{eventual_brokers.join(',')}" if @log_assignments
+        else
+          actual_reassignment += pr
           brokers = Scala::Collection::JavaConversions.as_java_iterable(replicas).to_a
-          @logger.info "Assigning #{topic_and_partition.topic},#{topic_and_partition.partition} to #{brokers.join(',')}"
+          @logger.debug "Assigning #{topic_and_partition.topic},#{topic_and_partition.partition} to #{brokers.join(',')}" if @log_assignments
         end
       end
       json = reassignment_json(actual_reassignment)
       @zk_client.reassign_partitions(json)
-      manage_overflow(reassignments)
+      manage_overflow(split(next_step_assignments, nil) + reassignments)
       manage_progress_state(actual_reassignment)
     end
 
-    private
+    def is_two_step_operation(topic_and_partition, final_replicas)
```

Comment:
This looks like it's a method that returns only `true` or `false`, but it seems to either compute something or return `false`. I would rename it after what it computes (would `#first_step_replicas` be a correct name?) and return `nil` in the case it shouldn't return anything.